contracts/interchain-token-service/src/state.rs

```diff
@@ -107,6 +140,89 @@ pub fn load_all_its_contracts(
         .collect::<Result<HashMap<_, _>, _>>()?)
 }
 
+pub fn save_token_info(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+    is_origin_chain: bool,
+    track_balance: bool,
+) -> Result<(), Error> {
+    match TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))? {
+        Some(_) if is_origin_chain => (),
+        Some(_) => return Err(Error::TokenNotDeployed { token_id, chain }),
+        None => {
+            let balance = TokenBalance {
+                balance: track_balance.then(Uint256::zero),
+                is_origin_chain,
+            };
+            TOKEN_INFO.save(storage, &(chain, token_id), &TokenInfo { balance })?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn update_token_info(
+    storage: &mut dyn Storage,
+    chain: ChainNameRaw,
+    token_id: TokenId,
+    amount: nonempty::Uint256,
+    is_deposit: bool,
```

Comment:
this flag is quite confusing, the way it is being passed around, and inverted, then x'or'd.